### PR TITLE
PERF: improved clip performance

### LINKED ISSUE #15400: PERF: Very slow clip performance
#### Code Sample, a copy-pastable example if possible

```python
In [38]: s = pd.Series(np.random.randn(30))

In [39]: timeit s.clip(0, 1)
100 loops, best of 3: 2.02 ms per loop
```

#### Problem description

There is more than 1000x performance difference between `Series.clip` and `numpy.clip`:

```
In [43]: timeit np.clip(arr, 0, 1)
1000000 loops, best of 3: 1.06 µs per loop
```

#### Output of ``pd.show_versions()``

pandas 0.19.2

### PR DESCRIPTION
**Pull request by jreback**
closes #15400
In [1]: np.random.seed(1234)

In [2]: s = pd.Series(np.random.randn(50))

master
```
In [3]: %timeit s.clip(0, 1)
1.65 ms ± 48.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

PR
```
In [3]: %timeit s.clip(0, 1)
124 µs ± 2.79 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

prob as good as can do for now as we still have 2 where ops (numpy does this in a single loop), and we have a mask check and fill (and final construction).

but about 15x better

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4104,6 +4105,22 @@ def isnull(self):
...
+    def _clip_with_scalar(self, lower, upper):
+
+        if ((lower is not None and np.any(isnull(lower))) or
+                (upper is not None and np.any(isnull(upper)))):
+            raise ValueError("Cannot use an NA value as a clip threshold")
+
+        result = self.values
+        mask = isnull(result)
+        if upper is not None:
+            result = np.where(result >= upper, upper, result)
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: I think this needs a `with np.errstate`, as we are working with raw array

```
In [8]: pd.Series([0, np.nan, 2]).clip(0, 1)
/home/joris/scipy/pandas/pandas/core/generic.py:4117: RuntimeWarning: invalid value encountered in greater_equal
  result = np.where(result >= upper, upper, result)
/home/joris/scipy/pandas/pandas/core/generic.py:4119: RuntimeWarning: invalid value encountered in less_equal
  result = np.where(result <= lower, lower, result)
Out[8]: 
0    0.0
1    NaN
2    1.0
dtype: float64
```

File: pandas/core/generic.py [hunk 2]
```diff
@@ -4104,6 +4105,22 @@ def isnull(self):
     def notnull(self):
         return notnull(self).__finalize__(self)
 
+    def _clip_with_scalar(self, lower, upper):
+
+        if ((lower is not None and np.any(isnull(lower))) or
+                (upper is not None and np.any(isnull(upper)))):
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: Are the `np.any` needed here? As lower/upper are already confirmed to be a scalar?

**Review by jorisvandenbossche — COMMENTED** [review]
There is a slight change in behaviour in that the new implementation does not preserve heterogeneous data types (eg int/float). 
Not that this should hold back these perf improvements, but we might consider to keep it for 0.21 (this is also no regression)

**Comment by jorisvandenbossche** [comment]
> new implementation does not preserve heterogeneous data types

In principle could add a check for that (at the if statement to decide to take this path or not), but not sure that is worth it ..

**Comment by jreback** [comment]
yeah this is a limitation of the current methodology

let me see what i can do

**Pull request merged**